crates/util/edcode-derive/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+                .into_compile_error()
+                .into());
+        }
+        let has_disc_err = !var.discriminant.clone().is_some_and(|(_, e)| {
+            enum_vals.push(e);
+            true
+        });
+        if has_disc_err {
+            return Err(Error::new(var.span(), discriminant_required!())
+                .into_compile_error()
+                .into());
+        }
+    }
+    let mut repr_type: Option<Ident> = None;
+    for attr in attrs {
+        if let Meta::List(meta) = attr.meta {
+            let is_repr = meta
+                .path
+                .require_ident()
+                .is_ok_and(|id| id == &Ident::new("repr", id.span()));
+            if is_repr {
+                let mut iter = meta.tokens.into_iter().peekable();
+                let span = iter.peek().span();
+                macro_rules! ident_helper {
+                            ($span:expr => $( $ty:ident ),*) => {
+                                vec![
+                                    $( Ident::new(stringify!($ty), $span) ),*
+                                ]
+                            };
+                        }
+                let supported = iter.next().is_some_and(|x| {
+                    if let TokenTree::Ident(id) = x {
+                        if ident_helper!(span => u8, u16, u32, u64, i8, i16, i32, i64).contains(&id)
+                        {
+                            repr_type = Some(id);
+                            true
+                        } else {
+                            false
+                        }
+                    } else {
+                        false
+                    }
+                });
+                if !supported {
+                    return Err(Error::new(span, unsupported_repr!())
+                        .into_compile_error()
+                        .into());
+                }
+            }
+        }
+    }
+    let repr_type = match repr_type {
+        None => {
+            return Err(Error::new(ident.span(), repr_required!())
+                .into_compile_error()
+                .into())
+        }
+        Some(x) => x,
+    };
+    Ok((ident, repr_type, enum_idents, enum_vals))
+}
+
+/// Derive `rimecraft_edcode::Encode` to objects.
+///
+/// # Enum
+///
+/// ## Requirements:
+/// - All variants must be field-less.
+/// - Enum must explicitly specify its representation through `#[repr()]`, and
+///   only `u*` and `i*` (excluding 128-bit types) repr are allowed.
+/// - All variants must have explicit discriminant.
+#[proc_macro_derive(Encode)]
+pub fn derive_encode(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    match input.data {
+        Data::Enum(data) => {
+            let (ident, repr_type, enum_idents, enum_vals) =
+                match parse_derive_enum(input.ident, input.attrs, data) {
+                    Ok(x) => x,
+                    Err(err) => return err,
+                };
```

Comment:
this could be replace by `parse_derive_enum(input.ident, input.attrs, data)?` (?